### ISSUE TITLE
Feature request: Switch GitLab release uploads to use Generic Packages Repository

### ISSUE THREAD

**Issue by sirlatrom** [goreleaser/goreleaser#2006]
**Is your feature request related to a problem? Please describe.**
With GitLab [Generic Packages Repository](https://docs.gitlab.com/ee/user/packages/generic_packages/#publish-a-generic-package-by-using-cicd) being available in Core/Free, it would be a better target for uploading binaries than the currently used [Upload Project File API endpoint](https://docs.gitlab.com/ce/api/projects.html#upload-a-file).

There are multiple benefits over "project files":
* The users can browse Generic Packages separately from and in addition to Releases in the GitLab UI
* Can still be used as release assets, as all you specify there is a URL
* API access for _downloading_ binaries is not supported for "project files", but is for Generic Packages

**Describe the solution you'd like**
Replace the upload method for GitLab releases to use the Generic Packages API, making sure the Release Asset URLs are correspondingly replaced.

**Describe alternatives you've considered**
Using the HTTP uploads feature to _additionally_ upload Generic Packages, as is possible with #2002 and #2005 combined. However, it requires additional configuration by the user, and results in two uploads per binary - one for "project files" and one for Generic Packages.

**Additional context**
GitLab documentation says the following about Generic Packages:

> Version history
> 
> - Introduced in GitLab 13.5.
> - It’s deployed behind a feature flag, enabled by default.
> - It’s enabled on GitLab.com.
> - It’s able to be enabled or disabled per-project.
> - It’s recommended for production use.
> - For GitLab self-managed instances, GitLab administrators can opt to disable it. 

- https://docs.gitlab.com/ee/user/packages/generic_packages/#publish-a-generic-package-by-using-cicd

**Comment by caarlos0** [comment]
cc/ @mavogel wdyt?

**Comment by mavogel** [comment]
IHMO it's a good idea because I faced trouble in the past with private `homebrew-taps` on gitlab and being able to download the binaries. I fixed it with a `PrivateRepositoryDownloadStrategy` and private session token from the browser, which was not my favorite way to go. If the Generic Packages Registry could solve this, I'd be delighted :)

**Comment by sirlatrom** [comment]
@mavogel You may want to read up on this GitLab issue regarding semantic version support (and the current lack thereof): https://gitlab.com/gitlab-org/gitlab/-/issues/255234

**Comment by mavogel** [comment]
Alright, thx for the update. As the PR shows it's part of the `13.11` release

**Comment by neomantra** [comment]
Publishing my project to Gitlab.com (hosted, but paid tiers), the process was failing with HTTP `413 Entity too large` errors.  Being new to goreleaser, I thought it was something with my configuration.  It took me way too long to realize that it was related to Gitlab.com's 10MB email attachment upload limit, which cannot be changed.

[I was also hit with the non-obvious "Gitlab job tokens don't work" of #2005 , took me a while to figure that out too (but my workaruond was to make another token).   Documenting that and making it more seamless could be its own issue.]

As a newcomer to `goreleaser`, it was quite rocky, but now I realize it is all GitLab quirks.  Implementing this would allow much more goreleaser'ing to Gitlab.com.

--- 

It seems that https://github.com/xanzy/go-gitlab/pull/1101 is in good shape... thanks for the progress.  What's the policy of putting support for this in `goreleaser` ahead of `xanzy/go-gitlab` merging that  `GenericPackagesService` PR?  

With respect to the `goreleaser` interface to this, I assume you would want backward compatibility since the Generic Packages Registry is not supported on older GitLab instances.    Would this require a new configuration flag?  Thoughts on this location:

```
gitlab_urls:
  # Use Generic Packages Repository for asset uploads, rather than email attachment
  use_generic_packages: true
```

**Comment by neomantra** [comment]
I got this into a works-for-me state.  There's two changesets, one for `go-gitlab` and one for `goreleaser`.  I didn't make PRs for either.

 * `go-gitlab`: https://github.com/neomantra/go-gitlab/commits/nm-1100
 * `goreleaser`: https://github.com/neomantra/goreleaser/commits/nm-2006

On the `goreleaser` side, I wasn't sure about using `NopCloser` versus anything else on this line:
https://github.com/neomantra/goreleaser/blob/nm-2006/internal/client/gitlab.go#L317

**Comment by caarlos0** [comment]
hey @neomantra , thanks for looking into this!

it looks good to me, should definitely open the PRs! 🙏

**Comment by neomantra** [comment]
I've started along updating for `go-gitlab`.  Will update the `goreleaser` side once that lands.  Thanks for the review!

**Comment by caarlos0** [comment]
Thanks @neomantra !

**Comment by neomantra** [comment]
It seems that [go-gitlab:#1100](https://github.com/neomantra/go-gitlab/commits/nm-1100) is moving along now.

I've rebased  the [nm-2006 branch](https://github.com/neomantra/goreleaser/tree/nm-2006) to the latest `goreleaser main` and updated it for changes to that  `go-gitlab` branch.   Once accepted in `go-gitlab`, I'll make a PR here....

**Comment by neomantra** [comment]
This feature has landed in `go-gitlab` so I made the PR.
